File: tests/test_stop.cpp
// The host's cancellation channel (#169): `SearchConfig::stop`,
// `ParallelConfig::stop`, and `TerminationReason::Cancelled`.
//
// Every test here is written the #104 way: it proves that the STOP is what
// ended the run, not that the run happened to be short. An iteration budget is
// set far above what the run can reach before the cancel, and the assertion is
// `termination == Cancelled` together with `iterations` strictly inside that
// budget -- so a stop check that was deleted would either run the budget out
// (failing on iterations) or report a different reason (failing on
// termination). None of them asserts on a duration.

#include <atomic>
#include <catch2/catch_test_macros.hpp>
#include <cbls/cbls.h>
#include <chrono>
#include <cstdint>
#include <memory>
#include <string>
#include <thread>
#include <vector>

using namespace cbls;

namespace {

// A continuous model with an objective, so the search can never exit on
// TerminationReason::Feasible and keeps finding new bests (the objective bound
// is tightened on each), which is what keeps the progress callback firing.
Model quadratic_model() {
    Model m;
    auto x = m.float_var(-5, 5);
    auto y = m.float_var(-5, 5);
    auto two = m.constant(2);
    auto neg1 = m.constant(-1.0);
    auto one = m.constant(1.0);
    m.add_constraint(m.sum({one, m.prod(neg1, x), m.prod(neg1, y)}));  // x + y >= 1
    m.minimize(m.sum({m.pow_expr(x, two), m.pow_expr(y, two)}));
    m.close();
    return m;
}

// Raise the token from inside the search's own progress callback rather than
// from a sleeping thread: the cancellation then lands at a deterministic point
// and the test needs no wall-clock assertion. tests/CMakeLists.txt's #104 note
// asks for exactly that restraint.
struct StopOnProgress : SolveCallback {
    StopToken* token;
    int seen = 0;
    explicit StopOnProgress(StopToken* t) : token(t) {}
    void on_progress(const SolveProgress& /*p*/) override {
        if (++seen >= 2) {
            token->request();
        }
    }
};

// Big enough that a search which ignored the stop would run visibly past the
// cancellation, small enough that such a regression fails in seconds rather
// than parking the suite.
constexpr int64_t kUnreachableIterations = 2000000;

}  // namespace

TEST_CASE("a host stop cancels a search that still has budget", "[stop]") {
    StopToken token;
    token.request();

    SearchConfig config;
    config.max_iterations = kUnreachableIterations;
    config.stop = token;

    Model m = quadratic_model();
    const SearchResult r =
        solve(m, /*time_limit=*/0.0, /*seed=*/9, true, nullptr, nullptr, 3, nullptr, config);

    // Not TimeLimit: this run had no wall clock at all. Not Stopped either --
    // no peer worker exists, and a host that cancelled needs to read back that
    // it was the cause.
    REQUIRE(r.termination == TerminationReason::Cancelled);
    REQUIRE(r.iterations < config.max_iterations);
}

TEST_CASE("a host stop raised mid-search ends the run where it stands", "[stop]") {
    // The distinguishing case: cancellation at a batch boundary part-way
    // through a search that is doing work. The test above pre-raises the flag,
    // so the loop condition is false on its first evaluation and no batch ever
    // runs.
    StopToken token;
    StopOnProgress cb(&token);

    SearchConfig config;
    config.max_iterations = kUnreachableIterations;
    config.batch_iterations = 100;
    config.stop = token;

    Model m = quadratic_model();
    const SearchResult r =
        solve(m, /*time_limit=*/0.0, /*seed=*/5, true, nullptr, nullptr, 3, &cb, config);

    REQUIRE(r.termination == TerminationReason::Cancelled);
    // Cancelled MID-FLIGHT, not before it started. Without this the test passes
    // on the pre-raised case too.
    REQUIRE(r.iterations > 0);
    REQUIRE(r.iterations < config.max_iterations);
}

TEST_CASE("an unattached stop leaves the run to its own budgets", "[stop]") {
    // The control for the two above. A `StopRef` that names nothing must never
    // read as requested, or every unconfigured run in the tree would end as
    // Cancelled -- which is the failure mode a default-constructed function
    // pointer invites.
    SearchConfig config;
    config.max_iterations = 5000;
    REQUIRE_FALSE(config.stop.attached());
    REQUIRE_FALSE(config.stop.requested());

    Model m = quadratic_model();
    const SearchResult r =
        solve(m, /*time_limit=*/0.0, /*seed=*/5, true, nullptr, nullptr, 3, nullptr, config);

    REQUIRE(r.termination == TerminationReason::IterationLimit);
    REQUIRE(r.iterations >= config.max_iterations);
}

TEST_CASE("an attached but unraised stop leaves the run to its own budgets", "[stop]") {
    // Stronger than the unattached control above, and the case that separates
    // `StopRef::requested()` from `StopRef::attached()` AT THE POLL SITE: a
    // `cancel_requested()` that asked whether a source is ATTACHED would end every
    // configured run as Cancelled, and every other solve in this file attaches a
    // token it then raises -- so nothing else here would notice.
    StopToken token;
    SearchConfig config;
    config.max_iterations = 5000;
    config.stop = token;
    REQUIRE(config.stop.attached());
    REQUIRE_FALSE(config.stop.requested());

    Model m = quadratic_model();
    const SearchResult r =
        solve(m, /*time_limit=*/0.0, /*seed=*/5, true, nullptr, nullptr, 3, nullptr, config);

    REQUIRE(r.termination == TerminationReason::IterationLimit);
    REQUIRE(r.iterations >= config.max_iterations);
    // And the search never wrote to the token.
    REQUIRE_FALSE(token.requested());
}

TEST_CASE("an attached stop is not a budget", "[stop]") {
    // Documented on `StopRef::attached()` and in docs/architecture.md, and worth
    // pinning because "run until the host cancels me" is the first thing a host
    // tries: with neither a wall clock nor an iteration limit the run returns at
    // once with NoBudget, attached stop or not. Treating an attachment as a budget
    // would hang forever the first time a host forgot to cancel.
    StopToken token;
    SearchConfig config;  // no max_iterations
    config.stop = token;

    Model m = quadratic_model();
    const SearchResult r =
        solve(m, /*time_limit=*/0.0, /*seed=*/5, true, nullptr, nullptr, 3, nullptr, config);

    REQUIRE(r.termination == TerminationReason::NoBudget);
    REQUIRE(r.iterations == 0);
}

TEST_CASE("a stop source is adapted by shape, not by type", "[stop]") {
    // StopRef takes anything with `bool requested() const`, which is what lets a
    // host hand over its own flag wrapper (or a C++20 std::stop_token) without
    // an adapter class.
    struct HostFlag {
        bool raised = false;
        [[nodiscard]] bool requested() const { return raised; }
    };

    HostFlag flag;
    const StopRef ref(flag);
    REQUIRE(ref.attached());
    REQUIRE_FALSE(ref.requested());
    flag.raised = true;
    REQUIRE(ref.requested());

    // And a default-constructed one is inert, with no source to read.
    const StopRef none;
    REQUIRE_FALSE(none.attached());
    REQUIRE_FALSE(none.requested());
}

TEST_CASE("a token can be reset and reused", "[stop]") {
    StopToken token;
    REQUIRE_FALSE(token.requested());
    token.request();
    REQUIRE(token.requested());
    token.reset();
    REQUIRE_FALSE(token.requested());
}

TEST_CASE("a host stop cancels every portfolio worker", "[stop][parallel]") {
    // ParallelConfig::stop reaches the workers through the combined StopRef in
    // solve_portfolio. The portfolio has NO wall clock here, so a worker that
    // ignored the cancel would grind its whole iteration budget and the
    // aggregate would come back IterationLimit.
    StopToken token;
    StopOnProgress cb(&token);

    SearchConfig config;
    config.max_iterations = kUnreachableIterations;
    config.batch_iterations = 100;

    ParallelConfig par_config;
    par_config.n_threads = 2;
    par_config.stop = token;

    ParallelSearch ps(2);
    const SearchResult r =
        ps.solve([] { return quadratic_model(); }, /*time_limit=*/0.0, /*seed=*/11, config,
                 /*hook_factory=*/nullptr, /*lns_factory=*/nullptr, &cb, par_config);

    REQUIRE(r.termination == TerminationReason::Cancelled);
    REQUIRE(r.iterations > 0);
    // Summed over both workers, so the budget to beat is twice one worker's.
    REQUIRE(r.iterations < 2 * config.max_iterations);
}

TEST_CASE("a stop raised from another thread reaches a running portfolio", "[stop][parallel]") {
    // The shape a host actually has: the cancel arrives from a thread that is
    // neither a worker nor the caller, while the solve is genuinely IN FLIGHT.
    //
    // "In flight" is the hard part, and an earlier version of this test did not
    // get it: opening the gate before `ps.solve` let the canceller fire before any
    // worker existed, so every worker broke out of its restart loop without ever
    // searching and the result came from the no-result fallback -- the same path
    // the pre-raised cases above already cover. The gate is now a PROGRESS ROW,
    // which only a running worker can produce, and `iterations > 0` is the
    // assertion that says the cancel landed on a search that had started.
    //
    // The wait is bounded so a regression cannot hang the suite: if no row ever
    // arrives the canceller fires anyway, inside the 30-second budget, so the run
    // is still mid-flight. That bound is not a duration assertion -- nothing here
    // asserts on elapsed time. The 30 seconds are there so that `Cancelled` is not
    // satisfiable by the clock expiring, which would report TimeLimit.
    struct GateOnProgress : SolveCallback {
        std::atomic<bool>* gate;
        explicit GateOnProgress(std::atomic<bool>* g) : gate(g) {}
        void on_progress(const SolveProgress& /*p*/) override {
            gate->store(true, std::memory_order_release);
        }
    };

    StopToken token;
    std::atomic<bool> searching{false};
    GateOnProgress cb(&searching);

    ParallelConfig par_config;
    par_config.n_threads = 2;
    par_config.stop = token;

    std::thread canceller([&token, &searching]() {
        const auto give_up = std::chrono::steady_clock::now() + std::chrono::seconds(5);
        while (!searching.load(std::memory_order_acquire) &&
               std::chrono::steady_clock::now() < give_up) {
            std::this_thread::yield();
        }
        token.request();
    });

    ParallelSearch ps(2);
    const SearchResult r =
        ps.solve([] { return quadratic_model(); }, /*time_limit=*/30.0, /*seed=*/13, SearchConfig{},
                 /*hook_factory=*/nullptr, /*lns_factory=*/nullptr, &cb, par_config);
    canceller.join();

    REQUIRE(r.termination == TerminationReason::Cancelled);
    REQUIRE(r.iterations > 0);
}

TEST_CASE("a cancel seen only after a worker's last solve still reports Cancelled",
          "[stop][parallel]") {
    // The window `with_host_cancel` exists for: a worker finishes a restart on its
    // ITERATION budget and the cancel is only visible afterwards. Reporting
    // IterationLimit there says the engine chose to stop, when the host took the
    // budget away.
    //
    // ONE worker, and that is load-bearing. With two, the first worker to finish
    // raises the token while its PEER is still mid-solve, the peer's own
    // `past_deadline()` sees it and returns Cancelled, and the aggregate is then
    // Cancelled for a completely different reason -- which is exactly how this test
    // passed with the guard neutered when it was first written. One worker has no
    // peer to see it.
    //
    // Reverting `with_host_cancel` to `return aggregate;` makes this report
    // IterationLimit.
    // A `Tracer`'s DESTRUCTOR runs on the worker's thread after its last solve and
    // before the portfolio aggregates, which is the only seam a host has into that
    // moment: callback, hook, LNS and tracer EVENTS all fire inside `cbls::solve`.
    struct RaiseOnDestruction : Tracer {
        StopToken* token;
        explicit RaiseOnDestruction(StopToken* t) : token(t) {}
        RaiseOnDestruction(const RaiseOnDestruction&) = delete;
        RaiseOnDestruction& operator=(const RaiseOnDestruction&) = delete;
        RaiseOnDestruction(RaiseOnDestruction&&) = delete;
        RaiseOnDestruction& operator=(RaiseOnDestruction&&) = delete;
        ~RaiseOnDestruction() override { token->request(); }
    };

    StopToken token;
    SearchConfig config;
    // No wall clock, so the worker runs exactly ONE solve and ends on this budget.
    config.max_iterations = 2000;
    config.batch_iterations = 100;

    ParallelConfig par_config;
    par_config.n_threads = 1;
    par_config.stop = token;
    par_config.tracer_factory = [&token](int /*worker*/) -> std::unique_ptr<Tracer> {
        return std::make_unique<RaiseOnDestruction>(&token);
    };

    ParallelSearch ps(1);
    const SearchResult r = ps.solve(
        [] { return quadratic_model(); }, /*time_limit=*/0.0, /*seed=*/17, config,
        /*hook_factory=*/nullptr, /*lns_factory=*/nullptr, /*callback=*/nullptr, par_config);

    REQUIRE(token.requested());
    // The worker did a full solve, so this is not the no-result fallback (which
    // reports Cancelled by a different route -- see empty_portfolio_reason).
    REQUIRE(r.iterations >= config.max_iterations);
    REQUIRE(r.termination == TerminationReason::Cancelled);
}

TEST_CASE("a wall clock that expires is not relabelled as a cancel", "[stop][parallel]") {
    // The other half of `with_host_cancel`, and why it is restricted to the
    // iteration budget: a portfolio whose SHARED CLOCK ran out is time-limited, and
    // a host cancel arriving in the same instant must not rewrite that. Same order
    // `ViolationLSLoop::run` applies by asking the clock first.
    //
    // One worker, for the same reason as the test above.
    // A `Tracer`'s DESTRUCTOR runs on the worker's thread after its last solve and
    // before the portfolio aggregates, which is the only seam a host has into that
    // moment: callback, hook, LNS and tracer EVENTS all fire inside `cbls::solve`.
    struct RaiseOnDestruction : Tracer {
        StopToken* token;
        explicit RaiseOnDestruction(StopToken* t) : token(t) {}
        RaiseOnDestruction(const RaiseOnDestruction&) = delete;
        RaiseOnDestruction& operator=(const RaiseOnDestruction&) = delete;
        RaiseOnDestruction(RaiseOnDestruction&&) = delete;
        RaiseOnDestruction& operator=(RaiseOnDestruction&&) = delete;
        ~RaiseOnDestruction() override { token->request(); }
    };

    StopToken token;
    ParallelConfig par_config;
    par_config.n_threads = 1;
    par_config.stop = token;
    par_config.tracer_factory = [&token](int /*worker*/) -> std::unique_ptr<Tracer> {
        return std::make_unique<RaiseOnDestruction>(&token);
    };

    ParallelSearch ps(1);
    const SearchResult r = ps.solve(
        [] { return quadratic_model(); }, /*time_limit=*/0.25, /*seed=*/19, SearchConfig{},
        /*hook_factory=*/nullptr, /*lns_factory=*/nullptr, /*callback=*/nullptr, par_config);

    REQUIRE(token.requested());
    REQUIRE(r.termination == TerminationReason::TimeLimit);
}

TEST_CASE("SearchConfig::stop and ParallelConfig::stop are OR-ed", "[stop][parallel]") {
    // Either channel alone cancels. A portfolio that read only one of them
    // would silently ignore whichever the caller happened to set.
    auto run = [](bool on_search_config, bool on_parallel_config) {
        StopToken token;
        token.request();
        SearchConfig config;
        config.max_iterations = kUnreachableIterations;
        if (on_search_config) {
            config.stop = token;
        }
        ParallelConfig par_config;
        par_config.n_threads = 1;
        if (on_parallel_config) {
            par_config.stop = token;
        }
        ParallelSearch ps(1);
        return ps
            .solve([] { return quadratic_model(); }, /*time_limit=*/0.0, /*seed=*/3, config,
                   nullptr, nullptr, nullptr, par_config)
            .termination;
    };

    REQUIRE(run(/*on_search_config=*/true, /*on_parallel_config=*/false) ==
            TerminationReason::Cancelled);
    REQUIRE(run(/*on_search_config=*/false, /*on_parallel_config=*/true) ==
            TerminationReason::Cancelled);
    REQUIRE(run(/*on_search_config=*/true, /*on_parallel_config=*/true) ==
            TerminationReason::Cancelled);
}

File: tests/python/test_stop.py
"""`cbls.StopToken`: cancelling a running solve from another Python thread (#169).

Every scenario here runs in a **child interpreter**, for the reason
`tests/python/test_pool.py` sets out at length and this file inherits: the
failure mode is a starved interpreter, and no in-process deadline can report
one. `cbls.solve` did not release the GIL before #169, so a Python thread that
wanted to call `StopToken.request()` mid-solve could not run at all -- the
regression this guards against is exactly a deadlocked child, where the solve
runs its whole budget while the canceller thread waits for a GIL it never gets.
`pytest-timeout` cannot see that (its `thread` method needs the same GIL; its
`signal` method only takes effect at the next bytecode in a main thread parked
inside a C++ call), so the deadline is a wall clock on a subprocess.

The C++ side of the same feature is `tests/test_stop.cpp`, which covers the
`TerminationReason::Cancelled` contract without an interpreter in the way.
"""

import os
import subprocess
import sys
import threading

import _cbls_core as cbls

# Generous against the ~1s solves below: this is a deadlock detector, not a
# performance floor.
CHILD_TIMEOUT_SECONDS = 30.0

# Far more GLS iterations than a cancelled run can reach, so "the stop ended it"
# is provable from the iteration count rather than from the wall clock -- the
# #104 discipline the C++ tests follow.
#
# Sized with a wide margin on purpose. `_scenario_cancel_running_solve` raises the
# token on a 0.5s timer, so a budget the search could EXHAUST inside that window
# would make the scenario report IterationLimit and fail with no other symptom --
# a flake. At this size an unpolled stop instead runs into the child's own
# wall-clock deadline, which `_assert_scenario_ok` reports by name.
UNREACHABLE_ITERATIONS = 200_000_000


def _quadratic() -> "cbls.Model":
    """x + y >= 1 over [-5, 5]^2, minimizing x^2 + y^2.

    An objective model, so the search never exits on `Feasible` and keeps
    tightening its bound instead of converging to a stop of its own.
    """
    m = cbls.Model()
    x = m.float_var(-5, 5)
    y = m.float_var(-5, 5)
    two = m.constant(2)
    neg1 = m.constant(-1.0)
    one = m.constant(1.0)
    m.add_constraint(m.sum([one, m.prod(neg1, x), m.prod(neg1, y)]))
    m.minimize(m.sum([m.pow_expr(x, two), m.pow_expr(y, two)]))
    m.close()
    return m


def _run_scenario(name: str) -> subprocess.CompletedProcess[str]:
    """Run one `__main__` scenario in a child interpreter, under a hard deadline."""
    env = dict(os.environ)
    # The child imports _cbls_core directly, without conftest.py to place the
    # build directory on the path -- hand it this process's search path instead.
    env["PYTHONPATH"] = os.pathsep.join(p for p in sys.path if p)
    # Every scenario assertion lives in the child, so an inherited
    # PYTHONOPTIMIZE would strip them and leave a child that prints OK and
    # exits 0 without having checked anything.
    env["PYTHONOPTIMIZE"] = "0"
    return subprocess.run(
        [sys.executable, os.path.abspath(__file__), name],
        capture_output=True,
        text=True,
        timeout=CHILD_TIMEOUT_SECONDS,
        check=False,
        env=env,
    )


def _assert_scenario_ok(name: str) -> str:
    try:
        proc = _run_scenario(name)
    except subprocess.TimeoutExpired as exc:
        raise AssertionError(
            f"scenario {name!r} did not finish within {CHILD_TIMEOUT_SECONDS}s -- "
            "cbls.solve is holding the GIL for the duration of the C++ call, so the "
            "thread that would raise the StopToken never runs (#169), or the stop is "
            "not being polled at all and the run is grinding its iteration budget"
        ) from exc
    assert proc.returncode == 0, f"scenario {name!r} failed:\n{proc.stdout}\n{proc.stderr}"
    return proc.stdout


def test_a_stop_token_cancels_a_running_solve() -> None:
    """A token raised from another Python thread ends `solve` with `Cancelled`.

    This is the whole feature: it needs the GIL release AND the stop poll, and
    it fails differently for each -- a hung child for the first, a
    `Cancelled`-less result for the second.
    """
    out = _assert_scenario_ok("cancel_running_solve")
    assert "OK" in out


def test_a_pre_raised_token_cancels_before_any_work() -> None:
    """The degenerate case, and the one that proves the reason is not a budget."""
    out = _assert_scenario_ok("cancel_before_start")
    assert "OK" in out


def test_an_unattached_stop_leaves_the_run_to_its_budget() -> None:
    """The control. Without it the two above pass on a solve that always cancels."""
    out = _assert_scenario_ok("no_stop")
    assert "OK" in out


def test_a_stop_token_cancels_a_portfolio() -> None:
    """`ParallelConfig.stop` reaches every worker."""
    out = _assert_scenario_ok("cancel_portfolio")
    assert "OK" in out


def test_stop_reads_back_as_attached() -> None:
    """The property round-trips as a bool, and None detaches.

    In-process: no thread and no solve, so nothing here can deadlock.
    """
    config = cbls.SearchConfig()
    assert config.stop is False
    token = cbls.StopToken()
    config.stop = token
    assert config.stop is True
    config.stop = None
    assert config.stop is False

    par = cbls.ParallelConfig()
    assert par.stop is False
    par.stop = token
    assert par.stop is True


def test_a_token_reports_and_resets_its_own_flag() -> None:
    """In-process: the token is a flag, and nothing about it needs a solve."""
    token = cbls.StopToken()
    assert not token.requested()
    token.request()
    assert token.requested()
    token.reset()
    assert not token.requested()


# --- child-process scenarios ------------------------------------------------


def _scenario_cancel_running_solve() -> None:
    token = cbls.StopToken()
    config = cbls.SearchConfig()
    config.max_iterations = UNREACHABLE_ITERATIONS
    config.stop = token

    # Raise the token shortly after the solve begins. A timer rather than a
    # sleep in the main thread: the main thread is inside the C++ call, which is
    # the point -- only a released GIL lets this one run at all.
    threading.Timer(0.5, token.request).start()

    result = cbls.solve(_quadratic(), 0.0, 7, config=config)
    assert result.termination == cbls.TerminationReason.Cancelled, result.termination
    assert 0 < result.iterations < UNREACHABLE_ITERATIONS, result.iterations


def _scenario_cancel_before_start() -> None:
    token = cbls.StopToken()
    token.request()
    config = cbls.SearchConfig()
    config.max_iterations = UNREACHABLE_ITERATIONS
    config.stop = token

    result = cbls.solve(_quadratic(), 0.0, 7, config=config)
    assert result.termination == cbls.TerminationReason.Cancelled, result.termination
    assert result.iterations < UNREACHABLE_ITERATIONS, result.iterations


def _scenario_no_stop() -> None:
    config = cbls.SearchConfig()
    config.max_iterations = 5000
    result = cbls.solve(_quadratic(), 0.0, 7, config=config)
    assert result.termination == cbls.TerminationReason.IterationLimit, result.termination
    assert result.iterations >= 5000, result.iterations


def _scenario_cancel_portfolio() -> None:
    token = cbls.StopToken()
    par = cbls.ParallelConfig()
    par.n_threads = 2
    par.stop = token

    threading.Timer(0.5, token.request).start()

    search = cbls.ParallelSearch(2)
    # A 60s clock the run must not reach: a portfolio that ignored the token
    # would come back TimeLimit a minute from now, or not at all.
    result = search.solve_parallel(
        _quadratic,
        60.0,
        11,
        cbls.SearchConfig(),
        None,
        None,
        None,
        par,
    )
    assert result.termination == cbls.TerminationReason.Cancelled, result.termination


if __name__ == "__main__":
    _scenarios = {
        "cancel_running_solve": _scenario_cancel_running_solve,
        "cancel_before_start": _scenario_cancel_before_start,
        "no_stop": _scenario_no_stop,
        "cancel_portfolio": _scenario_cancel_portfolio,
    }
    _scenarios[sys.argv[1]]()
    print("OK")
